tracer/src/Datadog.Trace/RuntimeMetrics/RuntimeMetricsWriter.cs

```diff
@@ -129,14 +130,31 @@ internal RuntimeMetricsWriter(IDogStatsd statsd, TimeSpan delay, bool inAzureApp
 
         public void Dispose()
         {
-            AppDomain.CurrentDomain.FirstChanceException -= FirstChanceException;
+            if (Interlocked.CompareExchange(ref _disposed, 1, 0) == 1)
+            {
+                Log.Debug("Disposing Runtime Metrics but it was already disposed");
+                return;
+            }
+
+            Log.Debug("Disposing Runtime Metrics");
             _timer.Dispose();
+            AppDomain.CurrentDomain.FirstChanceException -= FirstChanceException;
+            // We don't dispose runtime metrics on .NET Core because of https://github.com/dotnet/runtime/issues/103480
+#if NETFRAMEWORK
             _listener?.Dispose();
+#endif
             _exceptionCounts.Clear();
+            _statsd.Dispose();
         }
 
         internal void PushEvents()
         {
+            if (Volatile.Read(ref _disposed) == 1)
+            {
+                Log.Debug("Runtime metrics is disposed and can't push new events");
+                return;
+            }
```

Comment:
FYI, this doesn't make this thread-safe with the dispose, so there's still a race condition there 😄 Really the only way around it is a lock, but if we move the `IDogStatsd` instance disposal out of this type, then it's probably not a problem anyway 🙂